docs/developer-docs/latest/development/custom-fields.md

```diff
@@ -0,0 +1,340 @@
+    } 
+    options: {
+      base: [
+        /*
+          Declare settings to be added to the "Base settings" section
+          of the field in the Content-Type Builder
+        */ 
+        {
+          sectionTitle: { // Add a "Format" settings section
+            id: 'color-picker.color.section.format',
+            defaultMessage: 'Format',
+          },
+          items: [ // Add settings items to the section
+            {
+              /*
+                Add a "Color format" dropdown
+                to choose between 2 different format options
+                for the color value: hexadecimal or RGBA
+              */
+              intlLabel: {
+                id: 'color-picker.color.format.label',
+                defaultMessage: 'Color format',
+              },
+              name: 'options.format',
+              type: 'select',
+              value: 'hex', // option selected by default
+              options: [ // List all available "Color format" options
+                {
+                  key: 'hex',
+                  value: 'hex',
+                  metadatas: {
+                    intlLabel: {
+                      id: 'color-picker.color.format.hex',
+                      defaultMessage: 'Hexadecimal',
+                    },
+                  },
+                },
+                {
+                  key: 'rgba',
+                  value: 'rgba',
+                  metadatas: {
+                    intlLabel: {
+                      id: 'color-picker.color.format.rgba',
+                      defaultMessage: 'RGBA',
+                    },
+                  },
+                },
+              ],
+            },
+          ],
+        },
+      ],
+      advanced: [
+        /*
+          Declare settings to be added to the "Advanced settings" section
+          of the field in the Content-Type Builder
+        */ 
+      ],
+      validator: (args) => ({
+        'color-picker': yup.object().shape({
+          format: yup.string().oneOf(['hex', 'rgba']),
+        }),
+      }),
+    },
+  });
+}
+```
+
+:::
+
+::: note
+Relations, components or dynamic zones can't be used as a custom field's `type` parameter.
+:::
+
+### Components
+
+`app.customFields.register()` must pass a `components` object with 1 or 2 of the following components:
+
+| Component | Description                                                         |
+| --------- | ------------------------------------------------------------------- |
+| `Input`   | React component to use in the Content Manager's edit view           |
+| `View`    | Read-only React component to use in the Content Manager's list view |
+
+::: details Example: Registering Input and View components imported from dedicated files:
+
+<!-- ? can we declare components inline as well? like Input: async () => <MyReactComponent>…<MyReactComponent/>? What would be the most simple React.component example we could describe for a custom field? -->
```

Comment:
I don't think we should do that for 2 reasons:

- That would mean adding JSX to the file, we we'd need to `import React from 'react'` at the top. I think it's better to keep JSX in the component or pages folder where we expect them to be
- More importantly, in order to do [code splitting](https://reactjs.org/docs/code-splitting.html#code-splitting), we don't ask for the component directly, but for a async function that returns a promise with the component inside. I don't think the admin would support it if you provided the component directly

Reply:
Thanks for the detailed and helpful explanation 🤗